rpc: TestHeartbeatHealthTransport failed

rpc.TestHeartbeatHealthTransport [failed](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Nightlies_StressBazel/9641783?buildTab=log) with [artifacts](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Nightlies_StressBazel/9641783?buildTab=artifacts#/) on master @ [3382b2737ceb72de9b14d1d5923773718be5bca7](https://github.com/cockroachdb/cockroach/commits/3382b2737ceb72de9b14d1d5923773718be5bca7):


```
=== RUN   TestHeartbeatHealthTransport
I230417 10:57:12.434782 5077 2@rpc/context.go:2565  [rnode=1,raddr=127.0.0.1:41219,class=default,rpc] 21  connection is now ready
E230417 10:57:12.435540 5077 2@rpc/context.go:2388  [rnode=1,raddr=127.0.0.1:41219,class=default,rpc] 22  closing connection after: grpc: error reading from server: EOF [code 14/Unavailable]
I230417 10:57:12.435939 5077 rpc/context.go:2302  [rnode=1,raddr=127.0.0.1:41219,class=default,rpc] 23  connection heartbeat loop ended with err: grpc: error reading from server: EOF [code 14/Unavailable]
I230417 10:57:12.454973 4668 2@rpc/context.go:2565  [rnode=1,raddr=127.0.0.1:41219,class=default,rpc] 24  connection is now ready
E230417 10:57:12.455510 4668 2@rpc/context.go:2388  [rnode=1,raddr=127.0.0.1:41219,class=default,rpc] 25  closing connection after: grpc: connection error: desc = "transport: error while dialing: connection interrupted (did the remote node shut down or are there networking issues?)" [code 14/Unavailable]
I230417 10:57:12.455618 4668 rpc/context.go:2302  [rnode=1,raddr=127.0.0.1:41219,class=default,rpc] 26  connection heartbeat loop ended with err: grpc: connection error: desc = "transport: error while dialing: connection interrupted (did the remote node shut down or are there networking issues?)" [code 14/Unavailable]
E230417 10:57:12.456376 8654 2@rpc/context.go:2388  [rnode=1,raddr=127.0.0.1:41219,class=default,rpc] 27  unable to connect (is the peer up and reachable?): initial connection heartbeat failed: grpc: connection error: desc = "transport: error while dialing: dial tcp 127.0.0.1:41219: connect: connection refused" [code 14/Unavailable]
I230417 10:57:12.456477 8654 rpc/context.go:2302  [rnode=1,raddr=127.0.0.1:41219,class=default,rpc] 28  connection heartbeat loop ended with err: initial connection heartbeat failed: grpc: connection error: desc = "transport: error while dialing: dial tcp 127.0.0.1:41219: connect: connection refused" [code 14/Unavailable]
I230417 10:57:12.457821 8826 rpc/context.go:2302  [rnode=1,raddr=127.0.0.1:41219,class=default,rpc] 29  connection heartbeat loop ended with err: initial connection heartbeat failed: grpc: connection error: desc = "transport: error while dialing: dial tcp 127.0.0.1:41219: connect: connection refused" [code 14/Unavailable]
I230417 10:57:12.458728 9093 rpc/context.go:2302  [rnode=1,raddr=127.0.0.1:41219,class=default,rpc] 30  connection heartbeat loop ended with err: initial connection heartbeat failed: grpc: connection error: desc = "transport: error while dialing: dial tcp 127.0.0.1:41219: connect: connection refused" [code 14/Unavailable]
I230417 10:57:12.459482 9228 rpc/context.go:2302  [rnode=1,raddr=127.0.0.1:41219,class=default,rpc] 31  connection heartbeat loop ended with err: initial connection heartbeat failed: grpc: connection error: desc = "transport: error while dialing: dial tcp 127.0.0.1:41219: connect: connection refused" [code 14/Unavailable]
I230417 10:57:12.460283 9338 rpc/context.go:2302  [rnode=1,raddr=127.0.0.1:41219,class=default,rpc] 32  connection heartbeat loop ended with err: initial connection heartbeat failed: grpc: connection error: desc = "transport: error while dialing: dial tcp 127.0.0.1:41219: connect: connection refused" [code 14/Unavailable]
I230417 10:57:12.461090 9490 rpc/context.go:2302  [rnode=1,raddr=127.0.0.1:41219,class=default,rpc] 33  connection heartbeat loop ended with err: initial connection heartbeat failed: grpc: connection error: desc = "transport: error while dialing: dial tcp 127.0.0.1:41219: connect: connection refused" [code 14/Unavailable]
I230417 10:57:12.461784 9639 rpc/context.go:2302  [rnode=1,raddr=127.0.0.1:41219,class=default,rpc] 34  connection heartbeat loop ended with err: initial connection heartbeat failed: grpc: connection error: desc = "transport: error while dialing: dial tcp 127.0.0.1:41219: connect: connection refused" [code 14/Unavailable]
I230417 10:57:12.462468 9749 rpc/context.go:2302  [rnode=1,raddr=127.0.0.1:41219,class=default,rpc] 35  connection heartbeat loop ended with err: initial connection heartbeat failed: grpc: connection error: desc = "transport: error while dialing: dial tcp 127.0.0.1:41219: connect: connection refused" [code 14/Unavailable]
I230417 10:57:12.475000 9881 rpc/context.go:2302  [rnode=1,raddr=127.0.0.1:41219,class=default,rpc] 36  connection heartbeat loop ended with err: initial connection heartbeat failed: grpc: connection error: desc = "transport: error while dialing: dial tcp 127.0.0.1:41219: connect: connection refused" [code 14/Unavailable]
I230417 10:57:12.476426 11404 rpc/context.go:2302  [rnode=1,raddr=127.0.0.1:41219,class=default,rpc] 37  connection heartbeat loop ended with err: initial connection heartbeat failed: grpc: connection error: desc = "transport: error while dialing: dial tcp 127.0.0.1:41219: connect: connection refused" [code 14/Unavailable]
I230417 10:57:12.477470 11560 rpc/context.go:2302  [rnode=1,raddr=127.0.0.1:41219,class=default,rpc] 38  connection heartbeat loop ended with err: initial connection heartbeat failed: grpc: connection error: desc = "transport: error while dialing: dial tcp 127.0.0.1:41219: connect: connection refused" [code 14/Unavailable]
I230417 10:57:12.478802 11717 rpc/context.go:2302  [rnode=1,raddr=127.0.0.1:41219,class=default,rpc] 39  connection heartbeat loop ended with err: initial connection heartbeat failed: grpc: connection error: desc = "transport: error while dialing: dial tcp 127.0.0.1:41219: connect: connection refused" [code 14/Unavailable]
I230417 10:57:12.479549 11848 rpc/context.go:2302  [rnode=1,raddr=127.0.0.1:41219,class=default,rpc] 40  connection heartbeat loop ended with err: initial connection heartbeat failed: grpc: connection error: desc = "transport: error while dialing: dial tcp 127.0.0.1:41219: connect: connection refused" [code 14/Unavailable]
I230417 10:57:12.480322 11996 rpc/context.go:2302  [rnode=1,raddr=127.0.0.1:41219,class=default,rpc] 41  connection heartbeat loop ended with err: initial connection heartbeat failed: grpc: connection error: desc = "transport: error while dialing: dial tcp 127.0.0.1:41219: connect: connection refused" [code 14/Unavailable]
I230417 10:57:12.481206 12133 rpc/context.go:2302  [rnode=1,raddr=127.0.0.1:41219,class=default,rpc] 42  connection heartbeat loop ended with err: initial connection heartbeat failed: grpc: connection error: desc = "transport: error while dialing: dial tcp 127.0.0.1:41219: connect: connection refused" [code 14/Unavailable]
I230417 10:57:12.481993 12253 rpc/context.go:2302  [rnode=1,raddr=127.0.0.1:41219,class=default,rpc] 43  connection heartbeat loop ended with err: initial connection heartbeat failed: grpc: connection error: desc = "transport: error while dialing: dial tcp 127.0.0.1:41219: connect: connection refused" [code 14/Unavailable]
I230417 10:57:12.482714 12393 rpc/context.go:2302  [rnode=1,raddr=127.0.0.1:41219,class=default,rpc] 44  connection heartbeat loop ended with err: initial connection heartbeat failed: grpc: connection error: desc = "transport: error while dialing: dial tcp 127.0.0.1:41219: connect: connection refused" [code 14/Unavailable]
I230417 10:57:12.483475 12506 rpc/context.go:2302  [rnode=1,raddr=127.0.0.1:41219,class=default,rpc] 45  connection heartbeat loop ended with err: initial connection heartbeat failed: grpc: connection error: desc = "transport: error while dialing: dial tcp 127.0.0.1:41219: connect: connection refused" [code 14/Unavailable]
I230417 10:57:12.484117 12628 rpc/context.go:2302  [rnode=1,raddr=127.0.0.1:41219,class=default,rpc] 46  connection heartbeat loop ended with err: initial connection heartbeat failed: grpc: connection error: desc = "transport: error while dialing: dial tcp 127.0.0.1:41219: connect: connection refused" [code 14/Unavailable]
I230417 10:57:12.484739 12732 rpc/context.go:2302  [rnode=1,raddr=127.0.0.1:41219,class=default,rpc] 47  connection heartbeat loop ended with err: initial connection heartbeat failed: grpc: connection error: desc = "transport: error while dialing: dial tcp 127.0.0.1:41219: connect: connection refused" [code 14/Unavailable]
I230417 10:57:12.485596 12887 rpc/context.go:2302  [rnode=1,raddr=127.0.0.1:41219,class=default,rpc] 48  connection heartbeat loop ended with err: initial connection heartbeat failed: grpc: connection error: desc = "transport: error while dialing: dial tcp 127.0.0.1:41219: connect: connection refused" [code 14/Unavailable]
I230417 10:57:12.486263 13036 rpc/context.go:2302  [rnode=1,raddr=127.0.0.1:41219,class=default,rpc] 49  connection heartbeat loop ended with err: initial connection heartbeat failed: grpc: connection error: desc = "transport: error while dialing: dial tcp 127.0.0.1:41219: connect: connection refused" [code 14/Unavailable]
I230417 10:57:12.487040 13133 rpc/context.go:2302  [rnode=1,raddr=127.0.0.1:41219,class=default,rpc] 50  connection heartbeat loop ended with err: initial connection heartbeat failed: grpc: connection error: desc = "transport: error while dialing: dial tcp 127.0.0.1:41219: connect: connection refused" [code 14/Unavailable]
    context_test.go:1287: initial connection heartbeat failed: grpc: client requested node ID 1 doesn't match server node ID 123 [code 2/Unknown]
I230417 10:57:12.497828 13267 rpc/context.go:2302  [rnode=1,raddr=127.0.0.1:41219,class=default,rpc] 51  connection heartbeat loop ended with err: initial connection heartbeat failed: grpc: client requested node ID 1 doesn't match server node ID 123 [code 2/Unknown]
--- FAIL: TestHeartbeatHealthTransport (0.09s)
```
<p>Parameters: <code>TAGS=bazel,gss,deadlock</code>
</p>
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
/cc @cockroachdb/kv
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestHeartbeatHealthTransport.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)
</sub>


Jira issue: CRDB-27072

--- a/pkg/rpc/context_test.go
+++ b/pkg/rpc/context_test.go
@@ -1279,11 +1279,12 @@ func TestHeartbeatHealthTransport(t *testing.T) {
 	// happening for a while, but only comes out rarely when this package is
 	// stressed. This test is very aggressive since it is calling GRPCDialNode in
 	// a busy loop for 50ms.
-	expected := "doesn't match server cluster ID"
+	expectedCluster := "doesn't match server cluster ID"
+	expectedNode := "doesn't match server node ID"
 	// Should stay unhealthy despite reconnection attempts.
 	for then := timeutil.Now(); timeutil.Since(then) < 50*clientCtx.Config.RPCHeartbeatTimeout; {
 		err := clientCtx.TestingConnHealth(remoteAddr, serverNodeID)
-		if !isUnhealthy(err) && !testutils.IsError(err, expected) {
+		if !isUnhealthy(err) && !testutils.IsError(err, expectedCluster) && !testutils.IsError(err, expectedNode) {
 			t.Fatal(err)
 		}
 	}

